addons/knowledge/static/src/js/widgets/knowledge_emoji_picker.js

```diff
@@ -29,30 +27,116 @@ const EmojiPickerWidget = Widget.extend(WidgetAdapterMixin, {
+        this.el.style.display = 'none';
+    },
+    /**
+     * Show the emojiPicker and move it close to its target. Depending on the
+     * space available, it will be above or below the target, with its width
+     * to the right or to the left. The emojiPicker is displayed along with an
+     * invisible div which will force the user to close the emojiPicker before
+     * doing anything else. Listen to window scroll and resize events to replace
+     * the emojiPicker at a valid position.
+     *
+     * @param {Object} options
+     * @param {[HTMLElement]} currentTarget element which need an emoji change
+     *                        false => use the previous currentTarget
+     * @param {[boolean]} withTransition the emojiPicker will be moved to its
+     *                    target destination with an animation if True.
+     */
+    showOnTarget: function ({currentTarget=false, withTransition=false}) {
+        if (currentTarget) {
+            this.currentTarget = currentTarget;
+            // resize is needed for non-mobile devices
+            window.addEventListener('resize', this.windowChangeHandler);
+            // scroll is needed for mobile devices
+            window.addEventListener('scroll', this.windowChangeHandler);
+        } else if (!this.currentTarget) {
+            // undefined target, do nothing
+            return;
+        }
+        const container = this.currentTarget.ownerDocument.querySelector('.o_content');
+        const menu = this.el.querySelector('.o_article_emoji_dropdown_panel');
+        if (!container || !menu) {
+            // undefined configuration, do nothing
+            return;
+        }
+        this.el.style.display = 'block';
+        if (withTransition) {
+            menu.style.transition = 'top .3s ease 0s, left .3s ease 0s';
+        }
+        const targetRect = this.currentTarget.getBoundingClientRect();
+        const containerRect = container.getBoundingClientRect();
+        const menuRect = menu.getBoundingClientRect();
+        this.articleId = parseInt(this.currentTarget.dataset.articleId);
+        // picker below target
+        const topBelow = targetRect.top + targetRect.height;
+        // picker above target
+        const topAbove = targetRect.top - menuRect.height;
+        // picker left side aligned with target right side
+        const leftAligned = targetRect.left;
+        // picker right side aligned with target right side
+        const rightAligned = targetRect.left + targetRect.width - menuRect.width;
+        // if the bottom of the emojiPicker would be below the viewport and
+        // there is space to fully put it above
+        if (topBelow + menuRect.height > containerRect.top + containerRect.height &&
+            topAbove >= 0) {
+            menu.style.top = topAbove + 'px';
+        } else {
+            // default is below
+            menu.style.top = topBelow + 'px';
+        }
```

Comment:
Wow ... this seems like a lot of low level manipulations ... this makes code way too much complex. Isn't it achievable through standard css (with some mobile / display device css) ? Anyway this should be simplified a lot, normally JS should not have to do that kind of low level manipulation.

Reply:
It's not that complex, I don't think it is achievable through standard css because I use the same floating emojiPicker for the entire view, so its position in the view is not defined, and we need it alongside the button which opened it. There is already something similar in the editor for the floating toolbar.

https://github.com/odoo/odoo/blob/c7dee02324b13d7e192037f885f4527e2ab63d9d/addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js#L2064-L2125

The problem here is that the 2 html elements are unrelated in the view, so we have to position the emojipicker manually.

Before, it was not necessary, because we added an emojipicker clone for EVERY BUTTON that would open it, and it is worse than this computation in my opinion (because the picker was created inside the article handle and we could drag&drop the article by clicking on the emojipicker, also its position in the dom was before the following article handle so if its css position was set to relative, it would appear below the next handle, ...)

Though I'm always open to other ideas if you think something could work for this case